contracts/sg-ics721/src/testing/integration_tests.rs

```diff
@@ -91,68 +95,205 @@ fn no_init(
 ) {
 }
 
-#[derive(Debug)]
-struct Bech32AddressGenerator {
-    pub hrp: String,
+#[derive(Default)]
+pub struct MockAddressGenerator;
+
+impl AddressGenerator for MockAddressGenerator {
+    fn contract_address(
+        &self,
+        api: &dyn Api,
+        _storage: &mut dyn Storage,
+        code_id: u64,
+        instance_id: u64,
+    ) -> Result<Addr> {
+        let canonical_addr = Self::instantiate_address(code_id, instance_id);
+        Ok(Addr::unchecked(api.addr_humanize(&canonical_addr)?))
+    }
+
+    fn predictable_contract_address(
+        &self,
+        api: &dyn Api,
+        _storage: &mut dyn Storage,
+        _code_id: u64,
+        _instance_id: u64,
+        checksum: &[u8],
+        creator: &CanonicalAddr,
+        salt: &[u8],
+    ) -> Result<Addr> {
+        // string representation of the salt
+        let salt_to_string = std::str::from_utf8(salt)?;
+        // Remove the square brackets and split the string by commas
+        let parts: Vec<&str> = salt_to_string
+            .trim_matches(|c| c == '[' || c == ']')
+            .split(',')
+            .collect();
+        // Convert each part to a u8 and collect them into a Vec<u8>
+        let salt: Vec<u8> = parts
+            .iter()
+            .map(|part| part.trim().parse().unwrap())
+            .collect();
+        let canonical_addr = instantiate2_address(checksum, creator, &salt)?;
+        Ok(Addr::unchecked(api.addr_humanize(&canonical_addr)?))
+    }
 }
 
-const COUNT: Item<u8> = Item::new("count");
+impl MockAddressGenerator {
+    // non-predictable contract address generator, see `BuildContractAddressClassic`
+    // implementation in wasmd: https://github.com/CosmWasm/wasmd/blob/main/x/wasm/keeper/addresses.go#L35-L42
+    fn instantiate_address(code_id: u64, instance_id: u64) -> CanonicalAddr {
+        let mut key = Vec::<u8>::new();
+        key.extend_from_slice(b"wasm\0");
+        key.extend_from_slice(&code_id.to_be_bytes());
+        key.extend_from_slice(&instance_id.to_be_bytes());
+        let module = Sha256::digest("module".as_bytes());
+        let result = Sha256::new()
+            .chain(module)
+            .chain(key)
+            .finalize()
+            .to_vec()
+            .into();
+        return result;
+    }
+}
+pub struct MockApiBech32 {
+    prefix: &'static str,
+}
 
-impl Bech32AddressGenerator {
-    pub const fn new(hrp: String) -> Self {
-        Self { hrp }
+impl MockApiBech32 {
+    pub fn new(prefix: &'static str) -> Self {
+        Self { prefix }
     }
 }
 
-#[cw_serde]
-pub struct CustomClassData {
-    pub foo: Option<String>,
+impl Api for MockApiBech32 {
```

Comment:
Maybe could move all of these setup classes into a different file

Reply:
No, it is not as easy as it looks like, because this code is in sg-ics721 contract and also in ics721 package. So there will be duplicates in any case. Besides it is just test code, so here it is fine - in all other cases I wouldnt like duplicates neither :).

Reply:
Okay yeah, since its test code